Added scope organization param into MCP URL to allow specific organizations

Abeelha opened the pull request:
Issue in question: https://github.com/datopian/portaljs/issues/1477
---
Implemented changes in the MCP server to allow passing a organization name in the URL path, similar to the API endpoint, so that the MCP only returns results within that specific organization.
---

- The main API URL is dynamic now, depending if the user connector has an organization name such as:
- https://mcp.portaljs.com/sse ( it will work as normal having universal scope )
- https://mcp.portaljs.com/abeelha/sse ( now it will be limited to just have access to _"abeelha"_ organization data

>**NOTE:**
> removed .env since its not necessary _( also needs to remove the URL from the cloudflare worker variable enviroment )_

---

<!-- This is an auto-generated comment: release notes by coderabbit.ai -->
## Summary by CodeRabbit

* **New Features**
  * Organization-based routing: requests route to org-specific endpoints via path-based routes, including org-scoped realtime (SSE) streams.
  * Router derives org from URL and adjusts endpoint targets accordingly.

* **Bug Fixes**
  * Search includes query parameter only when provided.
  * Clear 404 responses for non-org or unmatched paths.

* **Chores**
  * Removed template entries from the example environment configuration.

<sub>✏️ Tip: You can customize this high-level summary in your review settings.</sub>
<!-- end of auto-generated comment: release notes by coderabbit.ai -->

Abeelha committed:
implementing url params to orgs such as "mcp.portaljs.com/@org-name URL pattern"

Abeelha committed:
Revert "implementing url params to orgs such as "mcp.portaljs.com/@org-name URL pattern""

This reverts commit 2797a3e6e4d89b63c5d8a57a885475b399649c9a.

Abeelha committed:
implemented url org api call ( needs refactoring and changes )

Abeelha committed:
Revert "implemented url org api call ( needs refactoring and changes )"

This reverts commit 5fe2453898bc4b18bf65552fbfa3b253ba713d5b.

Abeelha committed:
removed env URL API, now the MCP server URL/Param is being fetched and attached into portalJS api calls, scoping results from the organization stricly

Abeelha committed:
removed comments

Abeelha commented on src/index.ts:
```diff
@@ -2,20 +2,19 @@ import { McpAgent } from "agents/mcp";
+Global variable to store current org (set in router per request)
+or be default URL if not presented with the org name in the mcp url
+*/
+let CURRENT_API_URL = "https://api.cloud.portaljs.com";
```
this would break the scope URL-API-ORG implementation, and its not critical because PortalJS API is public

Abeelha committed:
remove debugging logs

Abeelha commented on src/index.ts:
```diff
@@ -342,14 +348,32 @@ export class MyMCP extends McpAgent<Env> {
+			}
+			if (endpoint === "mcp") {
+				return MyMCP.serve(`/${orgName}/mcp`).fetch(request, env, ctx);
+			}
```
<img width="408" height="37" alt="Code_VYVBTXdubZ" src="https://github.com/user-attachments/assets/da530652-ece0-4a33-b1f5-e852745a2257" />

---
**working** now with both | @ or " " | org names in URL params

Abeelha committed:
update regex to allow | @ or " " | org name in mcp URL

anuveyatsu commented on src/index.ts:
```diff
@@ -342,14 +348,33 @@ export class MyMCP extends McpAgent<Env> {
 		const url = new URL(request.url);
+		const pathname = url.pathname;
+
+		const orgMatch = pathname.match(/^\/(@)?([^\/]+)\/(sse|mcp)(\/.*)?$/);
```
@Abeelha I was wondering if we can't drop `/sse` from the path and only expect the following pattern:

`mcp.portaljs.com/@org-name`

then we would just don't expect any other variations. It must follow this pattern, otherwise, just fails.

Abeelha replied:
regarding the /sse in the URL, the SSE endpoint: The /sse part in the path (e.g., /@lcc/sse) is required by the MCP framework for Server-Sent Events transport, if we try to remove it, it may break MCP protocol

Abeelha replied:
tried doing a workaround but it doesnt work, mcp is not recognized as connector in claude desktop/gpt, etc.. if we remove the /sse from the URL

anuveyatsu commented on src/index.ts:
```diff
@@ -342,14 +348,33 @@ export class MyMCP extends McpAgent<Env> {
+
+		if (orgMatch) {
+			const [, hasAt, orgName, endpoint] = orgMatch;
+			const prefix = hasAt ? "@" : "";
```
so it must have `@` so we don't even check

Abeelha replied:
i didnt fully understand if you exactly wanted the "@" before the orgname, so i enabled both of them, making it possible to have both working, e.g: /lcc/sse - /@lcc/sse, i will remove the blank one and just keep with the "@" version

Abeelha replied:
now it just recognizes the @org-name , removed the other one

anuveyatsu commented on src/index.ts:
```diff
@@ -342,14 +348,33 @@ export class MyMCP extends McpAgent<Env> {
+			const [, hasAt, orgName, endpoint] = orgMatch;
+			const prefix = hasAt ? "@" : "";
+
+			CURRENT_API_URL = `https://api.cloud.portaljs.com/${prefix}${orgName}`;
```
so instead of doing this you should just replace the subdomain from `mcp.portaljs.com` to `api.cloud.portaljs.com` without parsing prefix, org name etc.

Abeelha replied:
i will be removing the prefix and orgname with a simple string replacement instead of parsing it so we can have a direct /@org-name/sse

anuveyatsu commented on src/index.ts:
```diff
@@ -342,14 +348,33 @@ export class MyMCP extends McpAgent<Env> {
+
+			CURRENT_API_URL = `https://api.cloud.portaljs.com/${prefix}${orgName}`;
+
+			if (endpoint === "sse") {
```
I know that Cf introduces `sse` in the path but is it necessary to have or can we drop it?

Abeelha replied:
it would need a work-around to remove the /sse or /mcp from the URL and may break our MCP

Abeelha replied:
tried making a workaround removing /see but its necessary from the call of MCP protocol
MyMCP.serveSSE("/sse").fetch(request, env, ctx);
it needs to have /sse and also the other params for it to function

anuveyatsu commented on src/index.ts:
```diff
@@ -342,14 +348,33 @@ export class MyMCP extends McpAgent<Env> {
+			if (endpoint === "sse") {
+				return MyMCP.serveSSE(`/${prefix}${orgName}/sse`).fetch(request, env, ctx);
+			}
+			if (endpoint === "mcp") {
```
I don't understand this, why do we have 2 options?

Abeelha replied:
currently we have both /sse and /mcp endpoints because the MCP framework supports different transports, i will remove the /mcp

Abeelha committed:
change pattern to have: @ - /@org-name/sse,
No parsing - just extract @org-name and use it directly,
Remove /mcp endpoint - only keep /sse,
Single CURRENT_API_URL definition at module level
removed regex aproach

Abeelha committed:
refactor code for better redability and documentation about it

Abeelha commented on src/index.ts:
```diff
@@ -2,20 +2,46 @@ import { McpAgent } from "agents/mcp";
 
-interface Env {
-	API_URL?: string;
+/**
+ * Base URL constant for PortalJS API
+ */
+const BASE_URL = "https://api.cloud.portaljs.com";
+
+/**
+ * Extract organization name from URL pathname
+ * Expected format: /@org-name/sse
+ * Returns: @org-name (e.g., "@lcc")
+ */
+function extractOrgFromPath(pathname: string): string | null {
+	if (pathname.startsWith('/@') && pathname.includes('/sse')) {
+		const sseIndex = pathname.indexOf('/sse');
+		return pathname.substring(1, sseIndex); // Extract @org-name
+	}
+	return null;
 }
```
The removal of the default /sse route is INTENTIONAL and matches the requirements:

  From Anuar's feedback in the meeting:
  "we should have just one way of doing this... we assume that it must be scoped... I don't see
  value... on entire portaljs instance"

  Current behavior is correct:
  - Only accepts: /@org-name/sse (e.g., /@lcc/sse)
  - Returns 404 for: /sse (unscoped)
  - This simplifies the code and enforces org-scoping;
  
  Removed Default Route (Lines 365)

  - Before: Had both /@org/sse AND /sse routes
  - After: Only /@org-name/sse route, returns 404 for anything else
  - Anuar's feedback: "we should have just one way of doing this... we assume that it must be
  scoped"

anuveyatsu reviewed (approved):
LGTM
